### PR TITLE
GP-2123 WDL workflow for bamQC

### PR DESCRIPTION
**Pull request by iainrb**
- Reimplementation of workflow in WDL
- https://jira.oicr.on.ca/browse/GP-2264 filtering input data
- Inline Python to evaluate downsampling parameters
- https://jira.oicr.on.ca/browse/GP-2094 Downsampling of input data to exact number of reads
- https://jira.oicr.on.ca/browse/GP-2291 Downsampled input to Picard MarkDuplicates
- Input to bam-qc-metrics 0.2.5:
  - Filtered and (if applicable) downsampled BAMs
  - Filtering results as metadata
- Updated Jenkins tests

### CODE REVIEW

**Review by mlaszloffy — APPROVED** [review]

File: filter_downsample.md [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+Downsampling thresholds are always evaluated on the BAM file _after_ filtering.
+
+### Methods
+
+We combine two ways of downsampling with samtools.
+
+#### Approximate downsampling
+
+This is done using `samtools view -s $FOO.$BAR`, where `$FOO` is a random seed and `$BAR` is a decimal expressing the probability of retaining a read. For example, to sample reads with a probability of 0.05 and random seed 99, use `samtools view -s 99.05`.
```
Comment (apmasell) [inline filter_downsample.md]: I would make these `$SEED` and `$PROB` or something to that effect. Also, this is the worst encoding abuse of a floating point I've seen

File: bamQC.wdl [hunk 2]
```diff
@@ -0,0 +1,758 @@
...
+        }
+        targets = {
+          "pre_ds": preDownsampleTarget,
+          "ds": downSampleTarget
+        }
+        statusFile = open("~{statusFile}", "w")
+        print(json.dumps(status), file=statusFile)
+        statusFile.close()
+        targetFile = open("~{targetsFile}", "w")
+        print(json.dumps(targets), file=targetFile)
```
Comment (apmasell) [inline bamQC.wdl]: Equivalent to `json.dump(targets, targetFile)`. There are some others. Probably not worth fixing, but in future...

**Review by apmasell — APPROVED** [review]

File: filter_downsample.md (hunk 1 above)
Comment (iainrb) [reply filter_downsample.md]: It is really quite awful. Probably Heng Li's fault.

File: bamQC.wdl (hunk 2 above)
Comment (iainrb) [reply bamQC.wdl]: Good to know. And easy enough to do a search and replace. Fixed.

**Pull request merged**